enterprise/reporting/error_reporting.go

```diff
@@ -0,0 +1,593 @@
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (edRep *ErrorDetailReporter) WaitForSetup(ctx context.Context, clientName string) error {
+	for {
+		if edRep.GetClient(clientName) != nil {
+			break
+		}
+		if err := misc.SleepCtx(ctx, time.Second); err != nil {
+			return fmt.Errorf("wait for setup: %w", ctx.Err())
+		}
+	}
+
+	return nil
+}
+
+func (*ErrorDetailReporter) IsPIIReportingDisabled(_ string) bool {
+	// Since we don't see the necessity for error detail reporting, we are implementing a kind of NOOP method
+	return false
+}
+
+func (edRep *ErrorDetailReporter) migrate(c types.Config) (*sql.DB, error) {
+	dbHandle, err := sql.Open("postgres", c.ConnInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &migrator.Migrator{
+		Handle:          dbHandle,
+		MigrationsTable: fmt.Sprintf("%v_migrations", ErrorDetailReportsTable),
+		// TODO: shall we use separate env ?
+		ShouldForceSetLowerVersion: config.GetBool("SQLMigrator.forceSetLowerVersion", true),
+	}
+	err = m.Migrate(ErrorDetailReportsTable)
+	if err != nil {
+		return nil, fmt.Errorf("could not run %v migrations: %w", ErrorDetailReportsTable, err)
+	}
+	return dbHandle, nil
+}
+
+func (edRep *ErrorDetailReporter) getWorkspaceID(sourceID string) string {
+	edRep.srcWspMutex.RLock()
+	defer func() { edRep.srcWspMutex.RUnlock() }()
+	return edRep.workspaceIDForSourceIDMap[sourceID]
+}
+
+func (edRep *ErrorDetailReporter) extractErrorDetails(sampleResponse string) errorDetails {
+	errMsg := edRep.errorDetailExtractor.GetErrorMessage(sampleResponse)
+	cleanedErrMsg := edRep.errorDetailExtractor.CleanUpErrorMessage(errMsg)
+	return errorDetails{
+		ErrorMessage: cleanedErrMsg,
+	}
+}
+
+func (edRep *ErrorDetailReporter) getDBHandle(clientName string) (*sql.DB, error) {
+	client := edRep.GetClient(clientName)
+	if client != nil {
+		return client.DbHandle, nil
+	}
+
+	return nil, fmt.Errorf("DBHandle not found for client name: %s", clientName)
+}
+
+func (edRep *ErrorDetailReporter) getTags(clientName string) stats.Tags {
+	return stats.Tags{
+		"workspaceID": edRep.workspaceID,
+		"namespace":   edRep.namespace,
+		"clientName":  clientName,
+		"instanceId":  edRep.instanceID,
+	}
+}
+
+// Sending metrics to Reporting service --- STARTS
+func (edRep *ErrorDetailReporter) mainLoop(ctx context.Context, clientName string) {
+	tags := edRep.getTags(clientName)
+
+	mainLoopTimer := stats.Default.NewTaggedStat("ed_main_loop_time", stats.TimerType, tags)
+	getReportsTimer := stats.Default.NewTaggedStat("ed_get_reports_time", stats.TimerType, tags)
+	getReportsCount := stats.Default.NewTaggedStat("ed_get_reports_count", stats.HistogramType, tags)
+	aggregateTimer := stats.Default.NewTaggedStat("ed_aggregate_time", stats.TimerType, tags)
+	getAggregatedReportsCount := stats.Default.NewTaggedStat("ed_aggregated_reports_count", stats.HistogramType, tags)
+
+	errorDetailReportsDeleteQueryTimer := stats.Default.NewTaggedStat("error_detail_reports_delete_query_time", stats.TimerType, tags)
```

Comment:
inconsistent metric naming:

```suggestion
	mainLoopTimer := stats.Default.NewTaggedStat("error_detail_reports_main_loop_time", stats.TimerType, tags)
	getReportsTimer := stats.Default.NewTaggedStat("error_detail_reports_get_time", stats.TimerType, tags)
	aggregateTimer := stats.Default.NewTaggedStat("error_detail_reports_aggregate_time", stats.TimerType, tags)
	getAggregatedReportsCount := stats.Default.NewTaggedStat("error_detail_reports_aggregated_count", stats.HistogramType, tags)

	errorDetailReportsDeleteQueryTimer := stats.Default.NewTaggedStat("error_detail_reports_delete_query_time", stats.TimerType, tags)
```

Reply:
No need for `getReportsCount` if you have `getReportsTimer`, timer as a histogram captures counts as well

Reply:
Can we do something like this ?
```go
getReportsTimer.Count(len(reports))
```

Reply:
@sanpj2292 I've miss-understood the intention of `getReportsCount := stats.Default.NewTaggedStat("ed_get_reports_count", stats.HistogramType, tags)` 

It needs to be kept, maybe renaming it to `getReportsSize` to avoid confusion with a counter.

Reply:
Put it back